dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusPTX.java

```diff
@@ -0,0 +1,310 @@
+import com.dat3m.dartagnan.expression.IValue;
+import com.dat3m.dartagnan.expression.op.IOpBin;
+import com.dat3m.dartagnan.parsers.LitmusPTXBaseVisitor;
+import com.dat3m.dartagnan.parsers.LitmusPTXParser;
+import com.dat3m.dartagnan.parsers.program.utils.AssertionHelper;
+import com.dat3m.dartagnan.parsers.program.utils.ProgramBuilder;
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.event.EventFactory;
+import com.dat3m.dartagnan.program.event.Tag;
+import com.dat3m.dartagnan.program.event.arch.ptx.RedOp;
+import com.dat3m.dartagnan.program.event.arch.ptx.AtomOp;
+import com.dat3m.dartagnan.program.event.core.Fence;
+import com.dat3m.dartagnan.program.event.core.Load;
+import com.dat3m.dartagnan.program.event.core.Store;
+import com.dat3m.dartagnan.program.memory.MemoryObject;
+import org.antlr.v4.runtime.misc.Interval;
+
+import java.math.BigInteger;
+
+import static com.dat3m.dartagnan.GlobalSettings.getArchPrecision;
+
+public class VisitorLitmusPTX extends LitmusPTXBaseVisitor<Object> {
+    private final ProgramBuilder programBuilder;
+    private int mainThread;
+    private int threadCount = 0;
+
+    public VisitorLitmusPTX(ProgramBuilder pb) {
+        this.programBuilder = pb;
+    }
+
+    // ----------------------------------------------------------------------------------------------------------------
+    // Entry point
+
+    @Override
+    public Object visitMain(LitmusPTXParser.MainContext ctx) {
+        visitThreadDeclaratorList(ctx.program().threadDeclaratorList());
+        visitVariableDeclaratorList(ctx.variableDeclaratorList());
+        visitInstructionList(ctx.program().instructionList());
+        if (ctx.assertionList() != null) {
+            int a = ctx.assertionList().getStart().getStartIndex();
+            int b = ctx.assertionList().getStop().getStopIndex();
+            String raw = ctx.assertionList().getStart().getInputStream().getText(new Interval(a, b));
+            programBuilder.setAssert(AssertionHelper.parseAssertionList(programBuilder, raw));
+        }
+        if (ctx.assertionFilter() != null) {
+            int a = ctx.assertionFilter().getStart().getStartIndex();
+            int b = ctx.assertionFilter().getStop().getStopIndex();
+            String raw = ctx.assertionFilter().getStart().getInputStream().getText(new Interval(a, b));
+            programBuilder.setAssertFilter(AssertionHelper.parseAssertionFilter(programBuilder, raw));
+        }
+        return programBuilder.build();
+    }
+
+
+    // ----------------------------------------------------------------------------------------------------------------
+    // Variable declarator list
+    @Override
+    public Object visitVariableDeclaratorLocation(LitmusPTXParser.VariableDeclaratorLocationContext ctx) {
+        programBuilder.initLocEqConst(ctx.location().getText(),
+                new IValue(new BigInteger(ctx.constant().getText()), getArchPrecision()));
```

Comment:
this can probably be replaced by `(IConst) ctx.constant().accept(this)`

Reply:
Replaced